Publish batching in Redis Engine

FZambia opened the pull request:
As part of #59

FZambia commented:
To summarize things done in this pr:

Added `pubRequest` struct and `pubCh` to Redis Engine. Now every publish results in sending `pubRequest` to `pubCh` - and one goroutine reads from that channel, batches publish requests and sends to Redis. Almost the same like @banks already did for managing subscriptions.

banks commented on libcentrifugo/api.go:
```diff
@@ -100,19 +101,29 @@ func (app *Application) broadcastCmd(cmd *broadcastAPICommand) (*response, error
-			return resp, nil
-		}
+		wg.Add(1)
+		go func(channel Channel) {
```
Hmm, this works but it feels off to me...

It's intentionally creating overhead and contention on the publish channel etc as well as spawning maybe thousands of goroutines. Perhaps that's not too bad in practice, but in theory it would be possible to do this much more efficiently if we had a real async API in our engine interface.

Maybe something for the future?

banks commented on libcentrifugo/engineredis.go:
```diff
@@ -613,28 +625,129 @@ func (e *RedisEngine) runPubSub() {
+}
+
+func fillPublishBatch(ch chan *pubRequest, prs *[]*pubRequest) {
+	for len(*prs) < RedisPublishBatchLimit {
```
Doesn't respect `RedisPublishBatchLimit` explicitly. For now I *guess* that's not a big deal because that limit is larger than buffer size, but I don't know what guarantees Go spec makes about this.

If another thread is pushing new things in constantly and the underlying implementation is a ring-buffer, presumably we could go round the ring many times and consume more than buffer size messages... Seems very unlikely in practice, but if we have a limit constant we should probably respect it explicitly.

banks commented on libcentrifugo/engineredis.go:
```diff
@@ -613,28 +625,129 @@ func (e *RedisEngine) runPubSub() {
+	}
+}
+
+func (e *RedisEngine) publish(chID ChannelID, message []byte, opts *publishOpts) error {
```
HOw about this very simple change to engine interface:

 - Make another method `publishAsync` which has same args but which returns `<-error` instead of `error`.
 - `publish` implementation then becomes:
```go
func (e *RedisEngine) publish(chID ChannelID, message []byte, opts *publishOpts) error {
    return <-e.publishAsync(chID, message, opts)
}
```
 - Cases like `broadcastCmd` above are now way more efficient and can just do something like:
```go
errs := make([]<-error, len(channels))
for i, channel := range channels {
    errs[i] = engine.publishAsync(channel, ...)
}
for _, err := range errs {
    if err != nil {
        // decide how to handle partial errors - just log?
    }
}

```

FZambia replied:
Btw I also thought about sth like this (but my idea was to change behaviour of `publish` to return channel so I rejected it as it results in massive refactoring inside other Centrifugo parts, and unfortunately I have not thought about possibility to simply add new method). Thanks, will try to change engine in this way!

FZambia commented:
Actually that was not so simple because we don't call engine method directly from `broadcastCmd` - we call application's `publish` method which contains lot of helper logic.

But anyway I think I changed it in a way you suggested - not adding new engine method though but returning error channel from existing and added helper method to publish synchronously to `Application`.

Result is wonderful: broadcast now 2 times faster for 10000 messages than version with goroutines.

FZambia commented:
Actually that was not so simple because we don't call engine method directly from `broadcastCmd` - we call application's `publish` method which contains lot of helper logic.

But anyway I think I changed it in a way you suggested - not adding new engine method though but returning error channel from existing and added helper method to publish synchronously to `Application`.

Result is wonderful: broadcast now 2 times faster for 10000 messages than version with goroutines.

banks commented:
Great!

Will look at code again tomorrow but that just didn't look very efficient!

Thanks

> On 22 Mar 2016, at 21:27, Alexandr Emelin <notifications@github.com> wrote:
> 
> Actually that was not so simple because we don't call engine method directly from broadcastCmd - we call application's publish method which contains lot of helper logic.
> 
> But anyway I think I changed it in a way you suggested - not adding new engine method though but returning error channel from existing and added helper method to publish synchronously to Application.
> 
> Result is wonderful: broadcast now 2 times faster for 10000 messages than version with goroutines.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

banks commented on libcentrifugo/application.go:
```diff
@@ -341,16 +341,11 @@ func (app *Application) pubControl(method string, params []byte) error {
-	adminChannel := app.config.AdminChannel
-	app.RUnlock()
-	return app.engine.publish(adminChannel, message, nil)
+func (app *Application) enginePublish(chID ChannelID, message []byte, opts *publishOpts) error {
```
Nitpick but that name is a little bit confusing. It's real purpose is a synchronous helper so maybe `publishSync` would be cleaner name?

Not a big deal if you disagree though.

FZambia replied:
Agree, just deleted it

banks commented:
This looks great to me :+1: 

If I get a chance, I'll try it out, but don't wait if you are happy.

FZambia commented:
ok, thanks a lot for looking! I can't find something I can improve here at moment - so going to merge this then. This async publish method in engine opens a road to further optimizations maybe - i.e. something like `async: true` in publish command if user don't need to wait for publish result to speed up publish commands coming from single Redis queue for example. But let's discuss this later outside this pr.
